src/evaluation/factuality.py

```diff
@@ -0,0 +1,289 @@
+from collections import defaultdict
+from typing import List
+from src.annotation import prompt_annotation_flow
+from src.detect_entities import detect_entities
+from src.entity_utils import MarkedEntity, count_entities, filter_entities
+from src.oracle import EntityMatchType, get_entity_annotations, oracle_label_entities
+from src.entity_factuality import ANNOTATION_LABELS
+from src.generation_utils import SUMMARY_FAILED_GENERATION
+from src.metrics import rouge
+import numpy as np
+from tqdm import tqdm
+from termcolor import colored
+
+
+def get_labeled_entities(
+    sums_by_id,
+    sum_ents_by_id,
+    gold_metadata,
+    xsum_test,
+    should_annotate,
+    entity_match_type,
+):
+    # Detect entities if they're not cached
+    if len(sum_ents_by_id) == 0:
+        for sum_id, summary in sums_by_id.items():
+            sum_ents_by_id[sum_id] = detect_entities(
+                summary, xsum_test[sum_id]["document"]
+            )
+    labeled_ents = oracle_label_entities(
+        sum_ents_by_id,
+        get_entity_annotations(sum_ents_by_id.keys(), gold_metadata),
+        entity_match_type,
+    )
+    unknown_entities = filter_entities(
+        lambda x: x["label"] == ANNOTATION_LABELS["Unknown"],
+        labeled_ents,
+    )
+
+    if should_annotate and count_entities(unknown_entities) > 0:
+        result = prompt_annotation_flow(
+            unknown_entities,
+            xsum_test,
+            sums_by_id,
+            gold_metadata,
+        )
+        if result is not False:
+            labeled_ents = oracle_label_entities(
+                sum_ents_by_id,
+                get_entity_annotations(sum_ents_by_id.keys(), gold_metadata),
+            )
+    return labeled_ents
+
+
+def mark_entities(summary, labeled_entities: List[MarkedEntity]):
+    marked_summary = summary
+    offset = 0
+    for ent in sorted(labeled_entities, key=lambda x: x["start"]):
+        position = ent["end"] + offset
+        inserted_label = f" [{ent['label']}]"
+        before = marked_summary[:position]
+        after = marked_summary[position:]
+        marked_summary = before + inserted_label + after
+        offset += len(inserted_label)
+    return marked_summary
+
+
+def evaluate_summary(
+    summary: str,
+    source: str,
```

Comment:
```suggestion
    source: str,
```

seems like arg is unused

Reply:
Leaving this since access to the source could likely be relevant for summary evaluation